tests/core/contracts/test_contract_constructor.py

```diff
@@ -184,13 +183,15 @@ def test_contract_constructor_build_transaction_with_constructor_with_argument(
         WithConstructorArgumentsContract,
         constructor_args,
         constructor_kwargs):
+    nonce = web3.eth.getTransactionCount(web3.eth.coinbase) + 1
```

Comment:
Is this supposed to  be an invalid nonce?  Otherwise, I *think* that the `+ 1` should be removed.

Reply:
If the nonce is 0, it will be changed to 1 in the return transaction. I am not sure why...

Reply:
Hrm, that sounds wrong, but likely a bug elsewhere (eth-tester)?.  Either way, it doesn't need to block this PR. 👍

Reply:
Another transaction is being sent in 187-188. Probably the most natural way to write this is to remove the `+ 1` and move this line down to just before 190.

Reply:
Of course! Stupid me :D

Reply:
The transaction nonce should always be the number of transactions the account has sent (prior to sending the current transaction).  So the first txn an account sends should have a nonce of `0`, second transaction `1`, and so on.

If you check the `nonce` against `web3.eth.getTransactionCount` *after* the transaction has been mined then it should satisfy `txn["nonce"] == web3.eth.getTransactionCount(addr) - 1`.

If the `nonce` value for a transaction is wrong (too large, too small), then the transaction should be rejected by the node.

Reply:
@pipermerriam thx for the explanation! Just one more question:D why do we want to specify nonce in transaction manually if it will be automatically assigned later anyway?